scheduler/utilization_based_host_allocator.go

```diff
@@ -0,0 +1,248 @@
+		"max_hosts": d.PoolSize,
+	}
+	avgMakespan := scheduledTasksDuration / time.Duration(d.PoolSize)
+	grip.AlertWhen(avgMakespan > dynamicDistroRuntimeAlertThreshold, underWaterAlert)
+
+	// log scheduler stats
+	grip.Info(message.Fields{
+		"message":                      "queue state report",
+		"runner":                       RunnerName,
+		"provider":                     d.Provider,
+		"distro":                       d.Id,
+		"pool_size":                    d.PoolSize,
+		"new_hosts_needed":             numNewHosts,
+		"num_existing_hosts":           len(existingHosts),
+		"num_free_hosts_approx":        numFreeHosts,
+		"queue_length":                 len(taskQueue),
+		"long_tasks":                   hostsForLongTasks,
+		"scheduled_tasks_runtime":      int64(scheduledTasksDuration),
+		"scheduled_tasks_runtime_span": scheduledTasksDuration.String(),
+	})
+
+	return numNewHosts, nil
+}
+
+// calcScheduledTasksDuration returns the total estimated duration of all
+// tasks scheduled to be run in a given task queue
+func calcScheduledTasksDuration(queue []model.TaskQueueItem) time.Duration {
+	var scheduledTasksDuration time.Duration
+
+	for _, taskQueueItem := range queue {
+		scheduledTasksDuration += taskQueueItem.ExpectedDuration
+	}
+	return scheduledTasksDuration
+}
+
+// calcNewHostsNeeded returns the number of new hosts needed based
+// on a heuristic that utilizes the total duration scheduled tasks,
+// targeting a maximum average duration for each task on all hosts
+func calcNewHostsNeeded(scheduledDuration, maxDurationPerHost time.Duration, numExistingHosts, numHostsNeededAlready int) int {
+	// number of hosts needed to meet the duration based turnaround requirement
+	// may be a decimal because we care about the difference between 0.5 and 0 hosts
+	numHostsForTurnaroundRequirement := float64(scheduledDuration) / float64(maxDurationPerHost)
+
+	// number of hosts that need to be spun up
+	numNewHostsNeeded := numHostsForTurnaroundRequirement - float64(numExistingHosts) + float64(numHostsNeededAlready)
+
+	// if we need less than 1 new host but have no existing hosts, return 1 host
+	// so that small queues are not stranded
+	if numExistingHosts < 1 && numNewHostsNeeded > 0 && numNewHostsNeeded < 1 {
+		return 1
+	}
+
+	// round the # of hosts needed down
+	numNewHosts := int(math.Floor(numNewHostsNeeded))
+
+	// return 0 if numNewHosts is less than 0
+	if numNewHosts < 0 {
+		numNewHosts = 0
+	}
+	return numNewHosts
+}
+
+// calcExistingFreeHosts returns the number of hosts that are not running a task,
+// plus hosts that will soon be free scaled by some fraction
+func calcExistingFreeHosts(existingHosts []host.Host, freeHostFactor float64, maxDurationPerHost time.Duration) (int, error) {
+	numFreeHosts := 0
+	if freeHostFactor > 1 {
+		return numFreeHosts, errors.New("free host factor cannot be greater than 1")
+	}
+
+	for _, existingHost := range existingHosts {
+		if existingHost.RunningTask == "" {
+			numFreeHosts++
+		}
+	}
```

Comment:
should we try and do this in an aggregation?

Reply:
I don't think so because this doesn't require a trip to the database. The list of hosts is already cached